mm2src/coins/z_coin/storage/walletdb/mod.rs

```diff
@@ -33,27 +45,93 @@ pub struct WalletDbShared {
+        height: block.height as u32,
+        hash,
+        time: block.time,
+        sapling_tree,
+    }))
+}
+
 #[cfg(not(target_arch = "wasm32"))]
 impl<'a> WalletDbShared {
-    pub async fn new(zcoin_builder: &ZCoinBuilder<'a>) -> MmResult<Self, WalletDbError> {
+    pub async fn new(zcoin_builder: &ZCoinBuilder<'a>, rpc: &mut impl ZRpcOps) -> MmResult<Self, WalletDbError> {
         let z_spending_key = match zcoin_builder.z_spending_key {
             Some(ref z_spending_key) => z_spending_key.clone(),
             None => extended_spending_key_from_protocol_info_and_policy(
                 &zcoin_builder.protocol_info,
                 &zcoin_builder.priv_key_policy,
             )
-            .map_err(|err| WalletDbError::ZCoinBuildError(err.to_string()))?,
+            .mm_err(|err| WalletDbError::ZCoinBuildError(err.to_string()))?,
+        };
+
+        let current_block_height = rpc
+            .get_block_height()
+            .await
+            .map_err(|err| WalletDbError::GrpcError(err.to_string()))?;
+
+        let sync_block = match zcoin_builder.z_coin_params.mode.clone() {
+            ZcoinRpcMode::Light { sync_params, .. } => {
+                let sync_height = match sync_params {
+                    Some(params) => match params {
+                        LightClientSyncParams::Date(date) => Some(
+                            zcoin_builder
+                                .calculate_starting_height_from_date(date, current_block_height)
+                                .map_err(|err| {
+                                    WalletDbError::ZcoinClientInitError(ZcoinClientInitError::ZcashDBError(err))
+                                })?,
+                        ),
+                        LightClientSyncParams::Height(height) => Some(height),
+                    },
+                    None => None,
+                };
+
+                match sync_height {
+                    Some(height) => checkpoint_block_from_height(height, rpc).await?,
+                    None => match zcoin_builder.protocol_info.check_point_block.clone() {
+                        Some(block) => Some(block),
+                        None => {
+                            let avg_blocktime = zcoin_builder.avg_blocktime().ok_or_else(|| {
+                                WalletDbError::ZcoinClientInitError(ZcoinClientInitError::ZcashDBError(format!(
+                                    "avg_blocktime not  specified in {} coin config",
+                                    zcoin_builder.ticker
+                                )))
+                            })?;
+                            let current_block_height = current_block_height - (avg_blocktime * 24);
+                            checkpoint_block_from_height(current_block_height, rpc).await?
+                        },
+                    },
+                }
+            },
+            ZcoinRpcMode::Native => zcoin_builder.protocol_info.check_point_block.clone(),
```

Comment:
Native should be the same as light client as discussed.

Reply:
This can be done in another PR as we agreed. Please don't forget to open an issue for it.